Add Tag component

CikiMomogi opened the pull request:
![Tag Preview](https://user-images.githubusercontent.com/74030149/193384090-a9cee2da-f40f-4d9b-849d-6af5d31e8dac.jpeg)
------------------------------------------

See the result : 

- <https://cikimomogi-grammy.github.io/PQ-507/hosting/deno-deploy.html>

- <https://cikimomogi-grammy.github.io/PQ-507/plugins/autoquote.html>

Some examples:

| Result                                                                                                               | Syntax                                                                                                                                             | Description                                                   |
| -------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------- |
| ![deno](https://user-images.githubusercontent.com/74030149/192101377-c0b9b7c9-bdb0-4d40-b3ab-65a3b3d61568.png)       | `<TagGroup><Tag template="deno"/></TagGroup>`                                                                                                          | Use default deno template                                     |
| ![denoland](https://user-images.githubusercontent.com/74030149/192101450-2576440e-abaf-4416-9aa2-8de642defddc.png)   | `<TagGroup><Tag template="deno" text="DENO LAND"/></TagGroup>`                                                                                         | Use default deno template but change the label to "DENO LAND" |
| ![denonodejs](https://user-images.githubusercontent.com/74030149/192101186-b990fa7d-91ab-4645-bb77-80101386d98f.png) | `<TagGroup><Tag template="deno"/><Tag template="nodejs"/></TagGroup>`                                                                                      | Use two tags: Deno and Node.JS                                |
| ![telegram](https://user-images.githubusercontent.com/74030149/192101520-3ea8d69e-41eb-4d8a-8c2c-5352cafce3eb.png)   | `<TagGroup><Tag text="GRAMMYJS" icon="telegram" iconType="logo" color="#26A5E4"/></TagGroup>`                                                      | Custom tag.                                                   |
| ![trending](https://user-images.githubusercontent.com/74030149/192101648-ce4389ac-451a-4f15-8ed8-6bac1800d88c.png)   | `<TagGroup><Tag text="TRENDING" icon="fire" iconType="icon" iconColor="#dc3545" iconBg="#ffe674" color="#fde427" textColor="#ff8080"/></TagGroup>` | Custom tag.                                                   |

------------------------------------------

List of `Tag` attributes:

<details>

| Attribute       | Option                                              | Description                                                                                                                                                                                                                                                                                |
| --------------- | --------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `template`          | `official`, `thirdparty`, `deno`, `nodejs`          | Tag template                                                                                                                                                                                                                                                                               |
| `desc`          | Any string                                          | Description for the tag. It will displayed as a tooltip                                                                                                                                                                                                                                    |
| `link`          | Any URL or anchor link.                             | Tag hyperlink                                                                                                                                                                                                                                                                              |
| `color`         | HEX, RGB, or any string that matched with css color | Tag background color. Example use for HEX color: `<Tag color="ffffff" ... />`                                                                                                                                                                                                              |
| `colorDark`     | HEX, RGB, or any string that matched with css color | Tag background color for dark mode. Example use for HEX color: `<Tag colorDark="ffffff" ... />`                                                                                                                                                                                            |
| `icon`          | .svg file's name                                    | Tag icon. If we do not have the icon on asset folder (matched by the file's name), then fetch them from [simpleicons.org](https://simpleicons.org). Example: `<Tag icon="npm" iconType="logo"/>` will fetch .svg with name `npm` from URL `https://simpleicons.org/icons/npm.svg`          |
| `iconType`      | `icon`, `logo`                                      | Icon type. `icon` will fetch from [icons.getbootstrap.com](https://icons.getbootstrap.com) and `logo` will fetch from [simpleicons.org](https://simpleicons.org). Example: `<Tag icon="fire" iconType="icon"/>` will fetch from URL `https://icons.getbootstrap.com/assets/icons/fire.svg` |
| `iconColor`     | HEX, RGB, or any string that matched with css color | Tag icon color                                                                                                                                                                                                                                                                             |
| `iconColorDark` | HEX, RGB, or any string that matched with css color | Tag icon color for dark mode                                                                                                                                                                                                                                                               |
| `iconBg`        | HEX, RGB, or any string that matched with css color | Background color of the tag icon                                                                                                                                                                                                                                                           |
| `iconBgDark`    | HEX, RGB, or any string that matched with css color | Background color of the tag icon for dark mode                                                                                                                                                                                                                                             |
| `text`          | Any string                                          | Tag label                                                                                                                                                                                                                                                                                  |
| `textColor`     | HEX, RGB, or any string that matched with css color | Tag label color. Example use for HEX color: `<Tag textColor="ffffff" ... />`                                                                                                                                                                                                               |
| `textColorDark` | HEX, RGB, or any string that matched with css color | Tag label color for dark mode. Example use for HEX color: `<Tag textColorDark="ffffff" ... />`                                                                                                                                                                                             |

</details>

------------------------------------------

`TagGroup` is useful for adding some margins to your tags. You can wrap your tags inside `TagGroup` like this one:

<details>

```jsx
<TagGroup>
  <Tag text="TAG 1" />
  <Tag text="TAG 2" />
</TagGroup>;
```

Always use `TagGroup` when `Tag` is used right below heading.
But, in some cases we may not include `TagGroup`, such as when we insert `Tag` between sentences like this one:

![inline-tag](https://user-images.githubusercontent.com/74030149/193386392-ca5b63fa-5db9-4035-9a7f-7cd07a3dd626.png)

List of `TagGroup` attributes:

| Attribute | Option                                                         | Description                                                      |
| --------- | -------------------------------------------------------------- | ---------------------------------------------------------------- |
| `mt`      | px, rem, or any string number that matched with css size units | Top margin. Example: `<TagGroup mt="12px"> ... </TagGroup>`      |
| `mb`      | px, rem, or any string number that matched with css size units | Bottom margin. Example: `<TagGroup mb="12rem"> ... </TagGroup>`  |
| `pt`      | px, rem, or any string number that matched with css size units | Top padding. Example: `<TagGroup pt="12rem"> ... </TagGroup>`    |
| `pb`      | px, rem, or any string number that matched with css size units | Bottom padding. Example: `<TagGroup pb="12rem"> ... </TagGroup>` |

</details>

KnorpelSenf commented:
Have you considered adding these labels to the navigation menu, too?

dcdunkan commented:
Suggestion: Maybe use the greenish color of Node.js logo for Node.js badge? (I don't see a specific color scheme for Deno other than black-and-white, so the current one is nice)

KnorpelSenf commented on site/docs/hosting/deno-deploy.md:
```diff
@@ -48,7 +48,7 @@ Here, we are using the bot token (`/<bot token>`).
 
 > This is the recommended method, and the easiest one to go with.
-> The main advantage of following this method is that Deno Deploy will watch for changes in your repository which includes your bot code, and it will deploy new versions automatically.
+> The main advantage of following method is that Deno Deploy will watch for changes in your repository which includes your bot code, and it will deploy new versions automatically.
```
The sentence was correct before, _following this method_ is what has the discussed advantage.

The sentence is incorrect now, if anything, it would have to be _of the following method_—but that's a very different statement then, and it's not what's meant here.

Please revert this one.

KnorpelSenf commented on site/docs/hosting/deno-deploy.md:
````diff
@@ -89,7 +89,7 @@ deployctl deploy --project <project> ./mod.ts --prod --token <token>
 To do that, send a request to
 
-```text
+```md
````
same as above

KnorpelSenf commented on site/docs/zh/hosting/deno-deploy.md:
````diff
@@ -89,7 +89,7 @@ deployctl deploy --project <project> ./mod.ts --prod --token <token>
 为了配置 webhook，发送一个请求到
 
-```text
+```md
````
same as above

CikiMomogi commented:
https://github.com/grammyjs/website/pull/507/commits/f06b4f76b8dd8741a2c7c3f46ab59cbcf5e7e8c2:

![chachacha](https://user-images.githubusercontent.com/74030149/191101740-33bcba27-3894-4d6f-8b73-4823e8744db3.png)

CikiMomogi commented:
To do:
- [ ] Implement labels into the navigation menu

KnorpelSenf commented:
As a general rule, all plugins should be available on both runtimes. That's why adding both Deno and Node labels to all official plugins is maybe a little weird. We should rather add a note to stateless-question (or finally port it, @EdJoPaTo?) as well as to fluent. It can say that exactly this plugin is Node-only.

We can well use the lables for the hosting pages!

CikiMomogi commented:
I had the same thought until I realized that some official plugins do not support Deno. Perhaps we can add a `node.js only` badge for those two plugins while removing the `Deno` and `Node.js` badges from the rest.

![nodejs-only](https://user-images.githubusercontent.com/74030149/191141821-c4efe09c-acac-49d5-8bdb-cc995c5a1ead.png)

KnorpelSenf commented:
I'm not sure we need a badge for that. That makes it look like Node-only plugins are a thing which we advertise or anything (we even created a label for it!) while in reality there's like 1 plugin that won't support Deno, and another one that will be supporting both runtimes in the future.

KnorpelSenf commented:
Great changes, awesome to use a component!

roj1512 reviewed (commented):
I have a doubt related to the SVG labels, specifically the ones saying "third-party":

I think that the current implementation will result in a small headache in the future for people willing to translate the website to another language, as they have to open the SVG in an editor, change the text, and then upload it to a different file.

Maybe we can make it a component and let it be usable, for example, this way:

```jsx
<Tag name="third-party" />
```

What do you think about it? I think it's going to be much much better to have less stuff for those who contribute new languages to take care of.

CikiMomogi commented:
> fetch came to Node v17 under experimental flag --experimental-fetch
> 
> It will be available in Node v18 without the flag.
> 
> https://github.com/nodejs/node/pull/41749#issue-1118239565
> 
> You no longer need any additional package to be installed

Hmm.. Let me try with https module.

CikiMomogi commented:
Some examples:

| Result                                                                                                               | Syntax                                                                                                                             | Description                                                                                                                           |
| -------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------- |
| ![deno](https://user-images.githubusercontent.com/74030149/192101377-c0b9b7c9-bdb0-4d40-b3ab-65a3b3d61568.png)       | `<TagGroup><Tag type="deno"/></TagGroup>`                                                                                          | Use the default deno template                                                                                                         |
| ![denoland](https://user-images.githubusercontent.com/74030149/192101450-2576440e-abaf-4416-9aa2-8de642defddc.png)   | `<TagGroup><Tag type="deno" text="DENO LAND"/></TagGroup>`                                                                         | Use the default deno template but change the label to "DENO LAND"                                                                     |
| ![denonodejs](https://user-images.githubusercontent.com/74030149/192101186-b990fa7d-91ab-4645-bb77-80101386d98f.png) | `<TagGroup><Tag type="deno"/><Tag type="nodejs"/></TagGroup>`                                                                      | Use two tags: Deno and Node.JS                                                                                                        |
| ![telegram](https://user-images.githubusercontent.com/74030149/192101520-3ea8d69e-41eb-4d8a-8c2c-5352cafce3eb.png)   | `<TagGroup><Tag logo="telegram" text="GRAMMYJS" color="#26A5E4"/></TagGroup>`                                                      | Custom tag.                                                                                                                           |
| ![trending](https://user-images.githubusercontent.com/74030149/192101648-ce4389ac-451a-4f15-8ed8-6bac1800d88c.png)   | `<TagGroup><Tag icon="fire" text="TRENDING" iconColor="#dc3545" iconBg="#ffe674" color="#fde427" textColor="#ff8080"/></TagGroup>` | Custom tag. If we do not have the icon on asset folder, then fetch them from [icons.getbootstrap.com](https://icons.getbootstrap.com) |


List of `<Tag />`'s attributes:

| Attribute   | Option                                                                                  | Description                                                                                                                                                                                                                                                                  |
| ----------- | --------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `type`      | `official`, `thirdparty`, `deno`, `nodejs`                                              | Tag template                                                                                                                                                                                                                                                                 |
| `color`     | HEX, RGB, or any string that matched with css color                                     | Tag background color. Example using HEX color: `<Tag color="ffffff"`                                                                                                                                                                                                         |
| `logo`      | The name of the .svg file from [simpleicons.org](https://simpleicons.org)               | Tag icon. If we do not have the icon on asset folder (matched by the file's name), then fetch them from [simpleicons.org](https://simpleicons.org). Example: `<Tag logo="npm"` will fetch the from URL `https://simpleicons.org/icons/npm.svg`                               |
| `icon`      | The name of the .svg file from [icons.getbootstrap.com](https://icons.getbootstrap.com) | Tag icon. If we do not have the icon on asset folder (matched by the file's name), then fetch them from [icons.getbootstrap.com](https://icons.getbootstrap.com). Example: `<Tag icon="fire"` will fetch the from URL `https://icons.getbootstrap.com/assets/icons/fire.svg` |
| `iconColor` | HEX, RGB, or any string that matched with css color                                     | Tag icon color                                                                                                                                                                                                                                                               |
| `iconBg`    | HEX, RGB, or any string that matched with css color                                     | Background color of the tag icon                                                                                                                                                                                                                                             |
| `text`      | Any string                                                                              | Tag label                                                                                                                                                                                                                                                                    |
| `textColor` | HEX, RGB, or any string that matched with css color                                     | Tag label color                                                                                                                                                                                                                                                              |
| `desc`      | Any string                                                                              | Description for the tag. It will displayed as a tooltip [*]                                                                                                                                                                                                                      |


`TagGroup` is useful for adding some margins to your tags. You can wrap your tags inside `TagGroup` like this one:

```jsx
<TagGroup>
  <Tag text="TAG 1"/>
  <Tag text="TAG 2"/>
</TagGroup>
```

List of `<TagGroup />`'s attributes:

| Attribute | Option                                                         | Description                    |
| --------- | -------------------------------------------------------------- | ------------------------------ |
| `mt`      | px, rem, or any string number that matched with css size units | Top margin. Example: `12px`    |
| `mb`      | px, rem, or any string number that matched with css size units | Bottom margin. Example: `12px` |

*Note: Tooltip feature is not ready yet. I'm working on it!

CikiMomogi commented:
This PR is ready for review. Just waiting for the last step to sync all languages once the PR is approved. I have updated some information too, please see the description above.

CikiMomogi commented:
Alright. This one is final. No more commit again from me. I promise :crossed_fingers:

KnorpelSenf commented:
As we're increasing the number of tools that are available to website contributors, we will have to start documenting how to document. I see two alternatives:

1. Just paste the PQ description and the end of the contribution guide into CONTRIBUTING.md. Pro: easy to write, con: unlikely anyone will read.
2. Start a page on https://grammy.dev itself about how to contribute. This could be useful for tips and tricks about OSS contributions in general, and it could contain instructions on how to help out with the grammY ecosystem as well as this site in particular. Pro: much easier to find (searchbar anyone?), con: takes some effort to write.

What is your opinion?

KnorpelSenf reviewed (changes requested):
These changes are great! Some sites have these ugly badges that just look like clutter. This is not that. I love how much effort you put into the design with the cute little animations and so on, it shows!

#### Plugin Badges in Nav Bar

I know I said that adding Deno/Node support badges to every official plugin page is superfluous (and I still think that). However, seeing now that the third-party plugins do have these labels (which makes sense), I must say that I find it a bit odd that they're missing from official plugins.

This leaves us with three choices:
1. Live with the inconsistency.
2. Add badges to all official plugins.
3. Remove badges from third-party plugins.

Personally, I think that the goal should be that _people do not have to think about the platform_. This just adds extra hassle. Ideally, I want that everyone can just use the tools they like, and everything will _just work_. I know that this is not quite the case, as there are a few edge cases where this does not work—however, in my opinion, I would not want to confront people with all the extra info, given that things work out of the box in 95 % of cases. Also, Node plugins can in fact often be used on Deno, so these badges are only correct to some extent.

One solution could be to leave out the badges everywhere in the nav bar for the list of plugins, but have them on each plugin page. That way, it is possible to view a consistent list of plugins (no badges anywhere) but as soon as you make a choice and want to read up about some plugin, the labels are still there, after the choice was made. So effectively we have a bit of option 3 and a bit of option 1.

In contrast, I think the badges work extremely well for the Hosting section, they just look stunning there and are super helpful!

What's your opinion on this? Naturally, if y'all disagree with me and there are other reasons I did not consider yet, feel free to raise your voice.

#### Tooling

When I clone this branch, there are a bunch of linting errors in my editor and warning from the Vetur extension. Could you add the list of extensions you use to the VSCode settings as extension recommendations? Also, I tend to lauch the editor at the repo root, but now it seems like VSCode does not pick up the right TS or Vue tooling when doing that. Seems like a lot broke, would appreciate if you could educate us.

CikiMomogi commented:
> As we're increasing the number of tools that are available to website contributors, we will have to start documenting how to document. I see two alternatives:
> 1. Just paste the PQ description at the end of the contribution guide into CONTRIBUTING.md. Pro: easy to write, con: unlikely anyone will read.
> 2. Start a page on https://grammy.dev itself about how to contribute. This could be useful for tips and tricks about OSS contributions in general, and it could contain instructions on how to help out with the grammY ecosystem as well as this site in particular. Pro: much easier to find (searchbar anyone?), con: takes some effort to write.
> 
> What is your opinion?


That's good idea. I don't mind creating a new page to explain more clearly the usage of this plugin.


> I know I said that adding Deno/Node support badges to every official plugin page is superfluous (and I still think that). However, seeing now that the third-party plugins do have these labels (which makes sense), I must say that I find it a bit odd that they're missing from official plugins.
> 
> This leaves us with three choices:
> 1. Live with the inconsistency.
> 2. Add badges to all official plugins.
> 3. Remove badges from third-party plugins.


I think number three is probably more reasonable. Thus, I have removed the Deno/Node tag from all plugin pages.


> When I clone this branch, there are a bunch of linting errors in my editor and warning from the Vetur extension. Could you add the list of extensions you use to the VSCode settings as extension recommendations? Also, I tend to lauch the editor at the repo root, but now it seems like VSCode does not pick up the right TS or Vue tooling when doing that. Seems like a lot broke, would appreciate if you could educate us.

 I read on the Vue page that Vetur has been deprecated and replaced by Volar. Here's the list of plugins I use:
- [TypeScript Vue Plugin (Volar)](https://marketplace.visualstudio.com/items?itemName=Vue.vscode-typescript-vue-plugin)
- [Vue Language Features (Volar)](https://marketplace.visualstudio.com/items?itemName=Vue.volar)
- [Prettier - Code formatter](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode)

I already set all the settings and include them in the recent commit. So, it should be possible to use both extensions immediately after installing them.

CikiMomogi commented on site/docs/.vuepress/public/tag/logo/nodedotjs.svg:
```diff
@@ -0,0 +1 @@
+<svg role="img" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><title>Node.js</title><path d="M11.998,24c-0.321,0-0.641-0.084-0.922-0.247l-2.936-1.737c-0.438-0.245-0.224-0.332-0.08-0.383 c0.585-0.203,0.703-0.25,1.328-0.604c0.065-0.037,0.151-0.023,0.218,0.017l2.256,1.339c0.082,0.045,0.197,0.045,0.272,0l8.795-5.076 c0.082-0.047,0.134-0.141,0.134-0.238V6.921c0-0.099-0.053-0.192-0.137-0.242l- ...
```
The Tag component works by requiring the user to specify the name of the icon file to be used. The component will then search for the file in the asset folder that we provide. If not found then download the file from the related source.

The name of the icon files inside the assets folder corresponds to the original name of the file from upstream. I can simply create an alias for "node" but the problem is if there is the exact same file name in the upstream (i.e. "node"), then the downloaded file will be different from what it should be.

KnorpelSenf replied:
That makes sense! Prefixing or postfixing would probably also work in order to improve consistency (`node-support` and `deno-support`) but I'm fine with this either way.

KnorpelSenf commented:
> > As we're increasing the number of tools that are available to website contributors, we will have to start documenting how to document. I see two alternatives:
> > 
> > 1. Just paste the PQ description at the end of the contribution guide into CONTRIBUTING.md. Pro: easy to write, con: unlikely anyone will read.
> > 2. Start a page on https://grammy.dev itself about how to contribute. This could be useful for tips and tricks about OSS contributions in general, and it could contain instructions on how to help out with the grammY ecosystem as well as this site in particular. Pro: much easier to find (searchbar anyone?), con: takes some effort to write.
> > 
> > What is your opinion?
> 
> That's good idea. I don't mind creating a new page to explain more clearly the usage of this plugin.

I have opened #519 to track and discuss this.

> > I know I said that adding Deno/Node support badges to every official plugin page is superfluous (and I still think that). However, seeing now that the third-party plugins do have these labels (which makes sense), I must say that I find it a bit odd that they're missing from official plugins.
> > This leaves us with three choices:
> > 
> > 1. Live with the inconsistency.
> > 2. Add badges to all official plugins.
> > 3. Remove badges from third-party plugins.
> 
> I think number three is probably more reasonable. Thus, I have removed the Deno/Node tag from all plugin pages.

Awesome, thanks! 🎉

> > When I clone this branch, there are a bunch of linting errors in my editor and warning from the Vetur extension. Could you add the list of extensions you use to the VSCode settings as extension recommendations? Also, I tend to lauch the editor at the repo root, but now it seems like VSCode does not pick up the right TS or Vue tooling when doing that. Seems like a lot broke, would appreciate if you could educate us.
> 
> I read on the Vue page that Vetur has been deprecated and replaced by Volar. Here's the list of plugins I use:
> 
> * [TypeScript Vue Plugin (Volar)](https://marketplace.visualstudio.com/items?itemName=Vue.vscode-typescript-vue-plugin)
> * [Vue Language Features (Volar)](https://marketplace.visualstudio.com/items?itemName=Vue.volar)
> * [Prettier - Code formatter](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode)
> 
> I already set all the settings and include them in the recent commit. So, it should be possible to use both extensions immediately after installing them.

Thank you so much! I'll try all of this out as soon as I have time. The plugins can also be added as recommended plugins to the `.vscode/settings.json` file.

CikiMomogi commented on site/docs/.vuepress/theme/components/AutoLink.vue:
```diff
@@ -0,0 +1,126 @@
+      <TagGroup v-if="item.favicon">
+        <Tag v-for="value in item.favicon" :item="value" />
+      </TagGroup>
+      <AutoLinkExternalIcon v-if="isBlankTarget" />
```
> was this code copied from the default theme? 
> If so, does that mean that we have to pull in changes from the upstream implementation periodically?

Yes, it was copied from the default theme. I've added comments on which parts of the code have been changed. Unfortunately, I couldn't find a more clever way than that.

KnorpelSenf replied:
Alright, thanks for the info. I'm fine with this solution.

CikiMomogi commented on site/docs/zh/plugins/transformer-throttler.md:
```diff
@@ -1,5 +1,7 @@
 # 流量控制（`transformer-throttler`）
 
+<TagGroup><Tag type="official" text="官方维护"/></TagGroup>
```
`Autotag` plugin to the rescue!

CikiMomogi commented:
Add two new plugins:
- `autotagPage` to add tags to multiple pages, and
- `autotagMenu` to add tags to multiple navbar menus.

I'm quite satisfied with the result. My code is probably a bit messy, so refactoring is very welcome :)
Once this PQ is successfully merged I will proceed with the documentation part.

KnorpelSenf commented:
Should we add these stunning looking badges to https://grammy.dev/hosting/comparison.html#serverless, as well?

CikiMomogi commented:
Sure. What badges should we add?

KnorpelSenf commented:
The Deno/Node ones.

Should we then retain the columns about Deno and Node? They will be redundant, but it could be that they still provide a better overview?

CikiMomogi commented:
Like this?

![A1](https://user-images.githubusercontent.com/74030149/195889402-6e6c7336-bddb-4bfa-b9a9-507ee83e6ff2.png)

Loskir commented:
No, I'd rather make a column "platforms"

CikiMomogi commented:
> No, I'd rather make a column "platforms"

And then we add Deno/Nodejs badges there?

Loskir commented:
Yes

CikiMomogi commented:
Style 1:

![A5](https://user-images.githubusercontent.com/74030149/195896867-89dea3e5-196b-4773-b4e4-78d16df0b033.png)

Style 2:
![A4](https://user-images.githubusercontent.com/74030149/195896951-d18826dc-7dde-4985-820c-07d8a16fd252.png)

Style 3: 
![A3](https://user-images.githubusercontent.com/74030149/195897108-6a72cdc0-71f7-4298-9cff-e2b683810e97.png)

Style 4: 
![A2](https://user-images.githubusercontent.com/74030149/195897239-0ed3fc0f-90f5-4dfd-a67b-64a0c81b2612.png)

CikiMomogi commented:
Personally, I prefer styles 1 or 2.

Loskir commented:
2 but a bit smaller paddings. 1 (current?) is also ok

KnorpelSenf commented:
Isn't web also a platform? Also, what does the web platform even tell us?

CikiMomogi commented:
> what does the web platform even tell us?

Good question :thinking:

Loskir commented:
> what does the web platform even tell us

ESM without any Deno APIs, I suppose

KnorpelSenf commented:
> > what does the web platform even tell us
> 
> ESM without any Deno APIs, I suppose

How about we add a Node badge if the platform supports Node APIs, a Deno badge if the platform supports Deno APIs, and no badge if the platform supports neither of the two?

Loskir commented:
Good, since the only platform that this is intended for was Cloudflare Workers, and it also supports Deno

CikiMomogi commented:
Do you think it would be a good idea to remove the pricing column? I believe that the `min. price` field already provides sufficient data, and that the addition of the price field only adds to the complexity.

KnorpelSenf commented:
Yes that sounds reasonable. I think the most important info to have on first sight is how cheap it could possibly get. Also, maintaining a list of constantly changing prices is extra work. We could save that.

CikiMomogi commented:
> Yes that sounds reasonable. I think the most important info to have on first sight is how cheap it could possibly get. Also, maintaining a list of constantly changing prices is extra work. We could save that.

Since some of the content needs to be translated, I've put them into different [PQ](https://github.com/grammyjs/website/pull/529). Now this PQ is ready to be reviewed.

KnorpelSenf commented:
This PQ causes the site's performance to drop heavily on pages that include a badge. For example, the total blocking time in a lighthouse test more than doubles from 100 ms to 230 ms. The badge is fading in after the page has loaded, which leads to a visible layout shift. This can be confirmed by looking at the cumulative layout shift value which increases by almost 50 %.
![image](https://user-images.githubusercontent.com/12952387/196497707-c0a2dee1-80bd-4415-9d74-0d596ceda4b0.png)

Other tools such as https://web.dev confirm these values.

In absolute values, I don't really think that we have crossed any thresholds the should prevent us from merging this. I am mainly posting this so that we all know what you're buying ourselves into.

The one thing that I believe does need improvement is the layout shift. It'd be great if the page just loads and that's it, rather than moving around the content after I started reading. @CikiMomogi what do you think?

CikiMomogi commented:
Hmm. Let's see if i can fix that.

CikiMomogi commented:
Closing this PQ temporarily. Perhaps I'll continue it in the future.

KnorpelSenf commented:
Sad to see this stalling :(
